PythonVisualizations/Heap.py

```diff
@@ -181,13 +184,212 @@ def siftUp(self, i):
+        callEnviron = self.createCallEnvironment()
+        if N is None:            # If N is not supplied,
+            N = len(self.list)   # then use number of items in list
+        heapLo = N // 2          # The heap lies in the range [heapLo, N)
+        while heapLo > 0:        # Heapify until the entire array is a heap
+            heapLo -= 1           # Decrement heap's lower boundary
+            self.siftDown(heapLo) # Sift down item at heapLo
+
+        self.cleanUp(callEnviron)
+              
     # lets user input an int argument that determines max size of the Heap
     def newArray(self):
         #gets rid of old elements in the list
         del self.list[:]
         self.size = 2
         self.display()
+    
+
+    def peek(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
 
+        # draw output box
+        canvasDimensions = self.widgetDimensions(self.canvas)
+        spacing = self.CELL_WIDTH * 3 // 4
+        padding = 10
+        boxSize = 1
+        peekIndex = 0
+        # changed the values that were added to the variables to get the output box in a desirable location
+        outputBox = self.canvas.create_rectangle(
+            (self.HEAP_X0 * 2 + padding),
+            canvasDimensions[1] - self.CELL_WIDTH - padding,
+            (self.HEAP_X0 * 2 + boxSize * spacing + padding * 2),
+            canvasDimensions[1] - padding,
+            fill=self.OPERATIONS_BG)
+        callEnviron.add(outputBox)
+
+        outputBoxCoords = self.canvas.coords(outputBox)
+        midOutputBox = (outputBoxCoords[3] + outputBoxCoords[1]) // 2
+        # create the value to move to output box
+        valueOutput = self.copyCanvasItem(self.list[peekIndex].display_val)
+        valueList = (valueOutput,)
+        callEnviron.add(valueOutput)
+        # move value to output box
+        # self.highlightCodeTags('print', callEnviron)
+        toPositions = (outputBoxCoords[0] + padding / 2 + (peekIndex + 1 / 2) * spacing,
+                        midOutputBox)
+        self.moveItemsTo(valueList, (toPositions,), sleepTime=.02)
+        # make the value 25% smaller
+        newFont = (self.VALUE_FONT[0], int(self.VALUE_FONT[1] * .75))
+        self.canvas.itemconfig(valueOutput, font=newFont)
+
+        # add label to output box
+        labelX = self.HEAP_X0 * 2 + boxSize * spacing + padding * 5
+        labelY = canvasDimensions[1] - padding * 2
+
+        boxLabel = self.canvas.create_text(
+            labelX, labelY, text="root", font=self.VARIABLE_FONT)
+        callEnviron.add(boxLabel)
+        self.highlightCodeTags([], callEnviron)
+        # finish animation
+        self.cleanUp(callEnviron)
+    
+    def removeMax(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+
+        # remove first element from list
+        n = self.list[0]
+        self.list[0] = None
+        self.size -= 1
```

Comment:
The textbook code for the Remove Max operation is:

```
   def remove(self):        # Remove top item of heap and return it
      if self.isEmpty():    # It's an error if the heap is empty
         raise Exception("Heap underflow")
      root = self._arr[0]   # Store the top item
      self._nItems -= 1     # Decrease item count
      self._arr[0] = self._arr[self._nItems] # Move last to root
      self._siftDown(0) # and then move it down into position
      return root           # Return top item
```
The test for the empty heap is handled by `clickRemoveMax()`.  The animation code should follow the textbook code closely.  The textbook does not store a `None` in the root location after copying that value to the `root` variable.  So line 332 should be removed.  Second, it decrements `nItems`, not the size of the array.  So line 333 should decrement `nItems` and the animation should show the pointer moving to its new location.